core/variant/struct.h

```diff
@@ -0,0 +1,91 @@
+/**************************************************************************/
+/*                         This file is part of:                          */
+/*                             GODOT ENGINE                               */
+/*                        https://godotengine.org                         */
+/**************************************************************************/
+/* Copyright (c) 2014-present Godot Engine contributors (see AUTHORS.md). */
+/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
+/*                                                                        */
+/* Permission is hereby granted, free of charge, to any person obtaining  */
+/* a copy of this software and associated documentation files (the        */
+/* "Software"), to deal in the Software without restriction, including    */
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#ifndef STRUCT_H
+#define STRUCT_H
+
+#include "core/string/string_name.h"
+#include "core/variant/variant.h"
+
+class Array;
+
+struct StructMember {
+	StringName name;
+	Variant::Type type; // TODO: should this be a union of type and class_name?
+	StringName class_name;
+
+	StructMember(const StringName &p_name, const Variant::Type p_type, const StringName &p_class_name = StringName()) {
+		name = p_name;
+		type = p_type;
+		class_name = p_class_name;
+	}
+};
+
+#define STRUCT_MEMBER(m_name, m_type) StructMember(SNAME(m_name), m_type)
+#define STRUCT_CLASS_MEMBER(m_name, m_class) StructMember(SNAME(m_name), Variant::OBJECT, m_class)
+// TODO: is there a way to define this so that the member count doesn't have to be passed?
+#define STRUCT_LAYOUT(m_name, m_member_count, ...)                               \
```

Comment:
I think passing member count is unnecesary, you can add some macro magic to guess the number of _VA_ARGS_.

Reply:
I couldn't figure out the proper way to guess the number of args. One method I saw worked using `sizeof`, but I don't think that will work here because `sizeof StructMember` won't be known yet? Another method involved explicitly enumerating all possible number of arguments up to some finite number (like 16, or 39, or whatever you have the patience to write). That method seemed pretty clunky to me, and would also limit the number of members a struct can have. That said, it would be great if the user didn't have to pass in the length somehow.

Reply:
You could check Boost.preprocessor for inspiration, there the macro would probably look more like (note lack of delimiting semicolons)
```
STRUCT_LAYOUT(PropertyInfoLayout,
		("name", Variant::STRING)
		("type", Variant::INT)
		("hint", Variant::INT)
		("hint_string", Variant::STRING)
		("class_name", Variant::STRING_NAME)
);
```
At which point `BOOST_PP_SEQ_SIZE` would give you the size. (and it's more generic in general, in case you ever needed a macro iterate over the member types more than once)
Or instead of getting size via a macro, you could put the field types into a tuple type and get tuple size at compile time via templates.

Reply:
Thanks for the tip! We won't to avoid unnecessary packages and libraries though, so tuple is off limits. A quick search on how `BOOST_PP_SEQ_SIZE` works suggests it falls in to the camp of explicitly enumerating sequence lengths up to some finite number. Maybe that's the way to go in the end, but it feels a little odd to restrict the number of members in a struct to a smallish number (like 64 or 256).

Reply:
> We won't to avoid unnecessary packages and libraries though, so tuple is off limits. 

Sure, but even without an `std::tuple`, it still can be done about the same, with the `sizeof...` operator. So as long as you can get the macro to generate `SomeTemplate<Variant::STRING, Variant::INT, Variant::INT, etc>`, that should be enough.

Reply:
Hmm, I'm not sure I fully understand what you're suggesting. Mind throwing an example my way? I'm still fairly new to C++ macros.

Reply:
Geez folks, please read my proposal, I posted a link on how to do this:
https://stackoverflow.com/questions/2124339/c-preprocessor-va-args-number-of-arguments

Reply:
Check the second answer

Reply:
Okay, I'll use that approach and limit the number of struct members to something like 64 or 128.

Reply:
I'll probably use a Python script or something to generate the macro up to whatever value we want. Should I include the python script as part of the Godot code base or just keep it local?

Alternatively, it seems like there might be a way to set up the macro and get parameter counting a bit more nicely using variadic templates. Is that an acceptable solution?

Reply:
I apologise if I'm missing something, but if you moved up the `static const StructMember members[member_count] =` line, couldn't you get the member count with the classic `sizeof(members)/sizeof(members[0])`?

Reply:
Uh, if you changed it from `static const StructMember members[member_count] =` to `static const StructMember members[] =`, that is.

Reply:
For some reason I thought that wouldn't work because StructMember isn't a primitive type (like int or something). I can try it out though, it would certainly be much simpler :)

Reply:
Actually, your idea seems to be working for me :)